clippy_lints/src/unwrap.rs

```diff
@@ -0,0 +1,176 @@
+                return unwrap_info;
+            },
+            _ => (),
+        }
+    } else if let Expr_::ExprUnary(UnNot, expr) = &expr.node {
+        return collect_unwrap_info(cx, expr, !invert);
+    } else {
+        if_chain! {
+            if let Expr_::ExprMethodCall(method_name, _, args) = &expr.node;
+            if let Expr_::ExprPath(QPath::Resolved(None, path)) = &args[0].node;
+            let ty = cx.tables.expr_ty(&args[0]);
+            if match_type(cx, ty, &paths::OPTION) || match_type(cx, ty, &paths::RESULT);
+            let name = method_name.name.as_str();
+            if ["is_some", "is_none", "is_ok", "is_err"].contains(&&*name);
+            then {
+                assert!(args.len() == 1);
+                let unwrappable = match name.as_ref() {
+                    "is_some" | "is_ok" => true,
+                    "is_err" | "is_none" => false,
+                    _ => unreachable!(),
+                };
+                let safe_to_unwrap = unwrappable != invert;
+                return vec![UnwrapInfo { ident: path, check: expr, safe_to_unwrap }];
+            }
+        }
+    }
+    Vec::new()
+}
+
+impl<'a, 'tcx: 'a> UnwrappableVariablesVisitor<'a, 'tcx> {
+    fn visit_branch(&mut self, cond: &'tcx Expr, branch: &'tcx Expr, else_branch: bool) {
+        let prev_len = self.unwrappables.len();
+        for unwrap_info in collect_unwrap_info(self.cx, cond, else_branch) {
+            if is_potentially_mutated(unwrap_info.ident, cond, self.cx)
+                || is_potentially_mutated(unwrap_info.ident, branch, self.cx)
+            {
+                // if the variable is mutated, we don't know whether it can be unwrapped:
+                continue;
+            }
+            self.unwrappables.push(unwrap_info);
+        }
+        walk_expr(self, branch);
+        self.unwrappables.truncate(prev_len);
+    }
+}
+
+impl<'a, 'tcx: 'a> Visitor<'tcx> for UnwrappableVariablesVisitor<'a, 'tcx> {
+    fn visit_expr(&mut self, expr: &'tcx Expr) {
+        if let Expr_::ExprIf(cond, then, els) = &expr.node {
+            walk_expr(self, cond);
+            self.visit_branch(cond, then, false);
+            if let Some(els) = els {
+                self.visit_branch(cond, els, true);
+            }
+        } else {
+            // find `unwrap[_err]()` calls:
+            if_chain! {
+                if let Expr_::ExprMethodCall(ref method_name, _, ref args) = expr.node;
+                if let Expr_::ExprPath(QPath::Resolved(None, ref path)) = args[0].node;
+                if ["unwrap", "unwrap_err"].contains(&&*method_name.name.as_str());
+                let call_to_unwrap = method_name.name == "unwrap";
+                if let Some(unwrappable) = self.unwrappables.iter()
+                    .find(|u| u.ident.def == path.def && call_to_unwrap == u.safe_to_unwrap);
+                then {
+                    self.cx.span_lint_note(
```

Comment:
clippy has various custom methods for linting, please use them. Most likely you'll want `span_lint_and_then`

Also don't use a `note`, but a `label`, as that will probably produce a prettier error message.

Reply:
Cool, `span_label` is much prettier :)